### DISCUSSION TITLE
Include an .env feature for dab.config.json file

### DISCUSSION THREAD

**Discussion by glaucia86** [Azure/data-api-builder#1361] [Ideas]
Hi! 

I would like to propose that we can make use of the .env file in the `dab.config.json` file.
As it already happens with the files generated by the SWA CLI: `staticwebapp.database.config.json` (example: **[HERE](https://github.com/Azure-Samples/dab-swa-library-demo/blob/main/swa-db-connections/staticwebapp.database.config.json#L8)**)

Because the connection string is very exposed, where it has extremely sensitive data such as: login, password and database name.

**Comment by Aniruddh25** [comment]
Hi @glaucia86! 
Thats a very good suggestion - in fact we support it already - the documentation needs to reflect this better. Here is the link on how to specify any value with an @env('ENV_NAME') - https://github.com/Azure/data-api-builder/blob/main/docs/configuration-file.md#accessing-environment-variables

Both `dab-config.json` and `staticwebapp.database.config.json` adhere to the same json config schema and are infact processed by the same dab service under the hood. So, you can provide the exact same contents in both the config files and they will be processed in the same way. In other words, they are just different names to the same config file based on the context you are using dab in.

Thanks,
Aniruddh!

**Reply by aaronpowell** [reply]
I think that @glaucia86's question is more around using the `.env` file (https://github.com/motdotla/dotenv) as a way to manage/load environment variables. The `.env` file is very common in Node.js apps especially, although there are implementations in most runtimes (https://www.nuget.org/packages/dotenv.net seems to be the most popular .NET implementation).

Admittedly the .NET approach is more to go down the route of using `appsettings.<env>.json` to load different settings, but DAB doesn't use the .NET config system and hierarchical configs are blocked by https://github.com/Azure/data-api-builder/issues/722

While it's true you can set system level environment variables and they use the `@env` loader in the DAB config, it's not a great DX as that makes it more difficult to do an isolated dev environment - imagine you're working on multiple projects using DAB, each with their own DB, how do you effectively manage all the related environment variables?

**Reply by glaucia86** [reply]
That's it! Exactly. See the image below:

[![Screen-Shot-03-27-23-at-08-27-AM.png](https://i.postimg.cc/HxGpQ6jT/Screen-Shot-03-27-23-at-08-27-AM.png)](https://postimg.cc/WDnLV6My)

I have a file called ".env."
As Aaron said, usually Node.js and TypeScript developers use the **dotenv package** a lot to manage their local variables in an application.

I created the `.env` file and followed the guidance that you gave me:

```json
{
   "connection-string": "@env('my-connection-string')"
}
```

But it didn't work. The documentation doesn't really make it very clear how to use it in this kind of circumstance.

> I'm also having Cross-Origin issues (using Axios in the Vue application). I will do new tests with @yorek today. If it doesn't work, I'll be opening a new issue here.

> if you need contact me, ping me on Teams

**Reply by Aniruddh25** [reply]
Would this `.env` file contain connection-string values which otherwise would be in the dab-config.json? How does that solve the problem of keeping the connection strings secure - we just moved the problem to a new file? I may be missing something here, haven't used .env file myself.

At the moment, the guidance to use `@env('local_machines_environment_variable_name_holding_value_for_my-connection-string')` is to provide it in the config file that you provide to dab. Its not expected to work if you provide that in a separate .env file.

**Reply by glaucia86** [reply]
And how to make it? Because the way it is in the documentation my connection string will be exposed (password, database name and other sensitive informations)

**Reply by aaronpowell** [reply]
@glaucia86 the `@env(...)` function would be used in the `dab.config.json` (or in your specific case `staticwebapp.database.json` file), not the `.env` file. Ultimately what happens is https://github.com/Azure/data-api-builder/blob/21df0a306a9df8b0645e4fad560b08f9f0a840be/src/Config/RuntimeConfigPath.cs#L38 is executed and if it finds the `@env` token it'll replace that with the value from `System.Environment.GetEnvironmentVariable` (see https://github.com/Azure/data-api-builder/blob/21df0a306a9df8b0645e4fad560b08f9f0a840be/src/Config/RuntimeConfigPath.cs#L124).

So you need to ensure that the variable is available there, either by setting it at a user/machine level, or exporting it from the shell before DAB starts.

@Aniruddh25 - yes, it does move the "problem" to another file, but this file is very common in Node.js projects (for which SWA is more targeted towards), and it's common to then have multiple files, with `.env` containing the common environment variables and `.env.<environment>` being used for environment specific ones (`.env.local` for local as an example). Often times projects will include a `.env.sample` as a "template" for what should be in the `.env` file, as you'll exclude them from git (to avoid secrets publishing).

It's the same approach as using `appsettings.json` to contain the non-environment-specific config but `appsettings.Development.json` for the stuff that's only applicable for the Development environment, such as a connection string. But as pointed out, we lack the ability to merge `dab.config.json` due to #722 so we need other ways to bootstrap environment variables.

**Reply by aaronpowell** [reply]
Given that this particular use case is with SWA Database Connections, would it make sense to raise it with the SWA CLI team on loading a `.env` file if it exists?

**Reply by aaronpowell** [reply]
I just realised that setting system-level (or user-level) environment variables in Windows has a barrier - it requires access that a user might not have in a managed environment. I've worked at enterprise organisations where you wouldn't be able to configure that.

**Reply by Aniruddh25** [reply]
I understand that setting environment variables at the user/system level is not possible for some users. 
Ability to merge multiple config files as per #722 could solve the `.env` issue as well. We are still looking if #722 can be unblocked (fyi: @abhishekkumams). We can revisit this discussion once that evaluation is complete. 

If #722 indeed is not possible, then we can look into using `dotenv.net` to read from a separate `.env` file - this might be useful for non-SWA users too.

**Reply by aaronpowell** [reply]
I'd actually love to revisit the config loading entirely and see if we can port it across to the standard .NET config system so that we don't have to build our parser.

**Reply by yorek** [reply]
I'm all in for supporting `.env` files. What I would like to have, ideally is

-  support for configuration files like .NET does (so one can have multiple configuration files for each environment, and they are merged together as described here: https://learn.microsoft.com/en-us/aspnet/core/fundamentals/configuration/?view=aspnetcore-7.0#appsettingsjson
- support for `.env` so that I can just put my sensitive data into the .env, if I don't need the complexity of having multiple configuration files.

By supporting both we'll be friendly both to MS (used to configuration files) and non-MS devs (used to .env)

**Reply by Aniruddh25** [reply]
Created this to track:
- #1374

**Reply by Aniruddh25** [reply]
This feature has been implemented and available in `0.8.44-rc`

**Comment by Aniruddh25** [comment]
This feature has been implemented and available in `0.8.44-rc`
